src/mailadm/conn.py

```diff
@@ -232,6 +232,48 @@ def get_expired_users(self, sysdate):
             users.append(UserInfo(*args))
         return users
 
+    def get_users_to_warn(self, sysdate: int) -> [{}]:
+        allusers = self.get_user_list()
+        users_to_warn = []
+        for user in allusers:
+            if user.token_name == "created in mailcow":
+                continue
+            year = 31536000
+            month = 2592000
+            week = 604800
+            day = 86400
+            warnmsg = "Your account will expire in ?. You should look for an alternative email " \
+                "provider right now.\nWith Delta Chat, you can keep all your chats and " \
+                "conversations. Just use the AEAP mechanism to tell your contacts of your " \
+                "address migration: https://delta.chat/en/2022-09-14-aeap\n\nYour %s team" % \
+                (self.config.mail_domain,)
+            if user.ttl >= year:
+                if user.warned == 0 and user.date + user.ttl < sysdate + month:
+                    users_to_warn.append({"user": user, "message": warnmsg.replace("?", "30 days")})
+                elif user.warned == 1 and user.date + user.ttl < sysdate + week:
+                    users_to_warn.append({"user": user, "message": warnmsg.replace("?", "7 days")})
+                elif user.warned == 2 and user.date + user.ttl < sysdate + day:
+                    users_to_warn.append({"user": user, "message": warnmsg.replace("?", "1 day")})
+            elif user.ttl >= month:
+                if user.warned == 0 and user.date + user.ttl < sysdate + week:
+                    users_to_warn.append({"user": user, "message": warnmsg.replace("?", "7 days")})
+                elif user.warned == 1 and user.date + user.ttl < sysdate + day:
+                    users_to_warn.append({"user": user, "message": warnmsg.replace("?", "1 day")})
+            elif user.ttl >= week:
+                if user.warned == 0 and user.date + user.ttl < sysdate + day:
+                    users_to_warn.append({"user": user, "message": warnmsg.replace("?", "1 day")})
+            else:
+                if user.warned == 0 and user.date + user.ttl < sysdate + user.ttl / 4:
+                    if user.ttl > day:
+                        timeleft = str(int(user.ttl / 4 / 60 / 60)) + " hours"
+                    else:
+                        timeleft = str(int(user.ttl / 4 / 60)) + " minutes"
+                    users_to_warn.append({"user": user, "message": warnmsg.replace("?", timeleft)})
```

Comment:
This looks like too much redundancy in the if-branches.  the `user` key in the appended dict is always the same so it can be pulled out of the branches. Also the branches can simply determine the replacement `warn_timing_message` text and at the end some `warnmsg.format(warn_timing_message)` is called once.

Reply:
good point, this is much nicer.